src/asm/lexer.cpp

```diff
@@ -77,17 +78,30 @@
 } while (0)
 #endif // !( defined(_MSC_VER) || defined(__MINGW32__) )
 
+struct CaseInsensitive {
+	// FNV-1a hash of an uppercased string
+	size_t operator()(std::string const &str) const {
+		size_t hash = 0x811C9DC5;
+
+		for (char const &c : str)
+			hash = (hash ^ toupper(c)) * 16777619;
+		return hash;
+	}
+
+	// Compare two strings without case-sensitivity (by converting to uppercase)
+	bool operator()(std::string const &str1, std::string const &str2) const {
+		return std::equal(RANGE(str1), RANGE(str2), [](char c1, char c2) {
+			return toupper(c1) == toupper(c2);
+		});
+	}
+};
+
 // Identifiers that are also keywords are listed here. This ONLY applies to ones
 // that would normally be matched as identifiers! Check out `yylex_NORMAL` to
 // see how this is used.
 // Tokens / keywords not handled here are handled in `yylex_NORMAL`'s switch.
-static struct KeywordMapping {
-	char const *name;
-	int token;
-} const keywords[] = {
-	// CAUTION when editing this: adding keywords will probably require extra nodes in the
-	// `keywordDict` array. If you forget to, you will probably trip up an assertion, anyways.
-	// Also, all entries in this array must be in uppercase for the dict to build correctly.
+// This assumes that no two keywords have the same name.
+static std::unordered_map<std::string, int, CaseInsensitive, CaseInsensitive> keywordDict = {
```

Comment:
I understand the desire to remove a bunch of LoC, especially this array that requires caring for `keywordDict` too; but OTOH I'm not convinced that this wouldn't penalise performance, since keywords are \*the\* most common token kind in RGBASM. (After whitespace, if you count that :P)

Reply:
I'll test the before+after code for performance and get your input on the results.

Reply:
The last commit to use the custom keyword trie was [`bc05437`](<https://github.com/gbdev/rgbds/pull/1313/commits/bc0543788955caed132f7ca7127cb4b1f4e6f351>):
- `make` builds a 247,248-byte `rgbasm`
- in `test/asm`,  `time ./test.sh` gives `real` time: [4.816s, 4.732s, 4.531s] = avg 4.693s
- in `pokecrystal` after `make tidy`, `time make RGBDS=../rgbds/ crystal` gives `real` time: [9.895s, 9.920s, 9.963s] = avg 9.926s

The commit that switched to `std::unordered_map` was [`ce56db2`](<https://github.com/gbdev/rgbds/pull/1313/commits/ce56db2e120c37c66fd113dbeaaa20f710555af4>):
- `make` builds a 252,344-byte `rgbasm` (+5,096 bytes)
- in `test/asm`,  `time ./test.sh` gives `real` time: [4.511s, 4.613s, 4.472s] = avg 4.532s (-0.161s)
- in `pokecrystal` after `make tidy`, `time make RGBDS=../rgbds/ crystal` gives `real` time: [10.090s, 10.212s, 10.043s] = avg 10.115s (+0.189s)

This is on par with all the other steady increases that using C++ STL has had, in cases where it's algorithmically identical (e.g. replacing a `realloc`ed array with `std::vector` has also added some "bloat").